### PR TITLE
Implemented Trustly

### PR DESCRIPTION
**Pull request by aait**

### CODE REVIEW

**Comment by payex-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetReversal.php  1
- src/SwedbankPay/Api/Service/Trustly/Resource/Request/Payment.php  2
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetSale.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetTransaction.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetTransactions.php  1
- tests/TrustlyPaymentTest.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/CreateSale.php  1
- src/SwedbankPay/Api/Service/Trustly/Resource/Request/PaymentObject.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetReversals.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetSales.php  1
- src/SwedbankPay/Api/Service/Trustly/Request/Purchase.php  1
- src/SwedbankPay/Api/Service/Trustly/Resource/Request/PaymentPayeeInfo.php  1
- src/SwedbankPay/Api/Service/Trustly/Request/GetPayment.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/CreateReversal.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/SwedbankPay/swedbank-pay-sdk-php/pullRequest?prid=6066095&bid=19866555)

**Comment by payex-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetTransaction.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetReversal.php  1
- src/SwedbankPay/Api/Service/Trustly/Resource/Request/Payment.php  2
- src/SwedbankPay/Api/Service/Trustly/Request/Purchase.php  1
- src/SwedbankPay/Api/Service/Trustly/Resource/Request/PaymentPayeeInfo.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetSale.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetTransactions.php  1
- src/SwedbankPay/Api/Service/Trustly/Resource/Request/PaymentObject.php  1
- tests/TrustlyPaymentTest.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetSales.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/CreateSale.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/CreateReversal.php  1
- src/SwedbankPay/Api/Service/Trustly/Request/GetPayment.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetReversals.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/SwedbankPay/swedbank-pay-sdk-php/pullRequest?prid=6066095&bid=19866555)

**Comment by payex-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/SwedbankPay/Api/Service/Trustly/Request/Purchase.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetReversal.php  1
- src/SwedbankPay/Api/Service/Trustly/Resource/Request/PaymentObject.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetTransactions.php  1
- tests/TrustlyPaymentTest.php  1
- src/SwedbankPay/Api/Service/Trustly/Resource/Request/Payment.php  2
- src/SwedbankPay/Api/Service/Trustly/Resource/Request/PaymentPayeeInfo.php  1
- src/SwedbankPay/Api/Service/Trustly/Request/GetPayment.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetSale.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/CreateReversal.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetTransaction.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetReversals.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetSales.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/CreateSale.php  1
         

Complexity decreasing per file
==============================
+ src/SwedbankPay/Api/Service/Creditcard/Resource/Request/PaymentUrl.php  -1
+ src/SwedbankPay/Api/Service/Paymentorder/Resource/PaymentorderUrl.php  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/SwedbankPay/swedbank-pay-sdk-php/pullRequest?prid=6066095&bid=19866555)

**Review by asbjornu — APPROVED** [review]

**Comment by asbjornu** [comment]
Please take a look at the failing checks.

**Comment by payex-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetSales.php  1
- src/SwedbankPay/Api/Service/Trustly/Request/GetPayment.php  1
- src/SwedbankPay/Api/Service/Trustly/Resource/Request/Payment.php  2
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetReversal.php  1
- src/SwedbankPay/Api/Service/Trustly/Request/Purchase.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/CreateReversal.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetReversals.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/CreateSale.php  1
- src/SwedbankPay/Api/Service/Trustly/Resource/Request/PaymentPayeeInfo.php  1
- src/SwedbankPay/Api/Service/Trustly/Resource/Request/PaymentObject.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetSale.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetTransaction.php  1
- tests/TrustlyPaymentTest.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetTransactions.php  1
         

Complexity decreasing per file
==============================
+ src/SwedbankPay/Api/Service/Paymentorder/Resource/PaymentorderUrl.php  -1
+ src/SwedbankPay/Api/Service/Creditcard/Resource/Request/PaymentUrl.php  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/SwedbankPay/swedbank-pay-sdk-php/pullRequest?prid=6066095&bid=19866555)

**Review by asbjornu — APPROVED** [review]

**Comment by payex-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- src/SwedbankPay/Api/Service/Trustly/Resource/Request/PaymentObject.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetTransaction.php  1
- src/SwedbankPay/Api/Service/Trustly/Request/GetPayment.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/CreateReversal.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/CreateSale.php  1
- tests/TrustlyPaymentTest.php  2
- src/SwedbankPay/Api/Service/Trustly/Transaction/Resource/Response/Sale.php  1
- src/SwedbankPay/Api/Service/Trustly/Resource/Request/PaymentPayeeInfo.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetTransactions.php  1
- src/SwedbankPay/Api/Service/Trustly/Resource/Request/Payment.php  2
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetReversals.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetSales.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetSale.php  1
- src/SwedbankPay/Api/Service/Trustly/Request/Purchase.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetReversal.php  1
         

Complexity decreasing per file
==============================
+ src/SwedbankPay/Api/Service/Creditcard/Resource/Request/PaymentUrl.php  -1
+ src/SwedbankPay/Api/Service/Paymentorder/Resource/PaymentorderUrl.php  -1
         

Clones added
============
- src/SwedbankPay/Api/Service/Trustly/Transaction/Resource/Response/Sale.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/SwedbankPay/swedbank-pay-sdk-php/pullRequest?prid=6066095&bid=19866555)

File: tests/TrustlyPaymentTest.php [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+use SwedbankPay\Api\Service\Trustly\Transaction\Request\GetSales;
+use SwedbankPay\Api\Service\Trustly\Transaction\Request\GetReversals;
+use SwedbankPay\Api\Service\Trustly\Transaction\Request\GetTransactions;
+
+use SwedbankPay\Api\Service\Payment\Transaction\Resource\Response\SalesObject;
+use SwedbankPay\Api\Service\Payment\Transaction\Resource\Response\ReversalsObject;
+use SwedbankPay\Api\Service\Payment\Transaction\Resource\Response\TransactionsObject;
+// phpcs:enable
+
+class TrustlyPaymentTest extends TestCase
```
Comment (payex-dev) [inline tests/TrustlyPaymentTest.php:37]: Codacy found an issue: [The class TrustlyPaymentTest has a coupling between objects value of 26. Consider to reduce the number of dependencies under 13.](https://app.codacy.com/gh/SwedbankPay/swedbank-pay-sdk-php/pullRequest?prid=6066095)

**Comment by payex-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetReversal.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Resource/Response/Sale.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetTransactions.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetReversals.php  1
- src/SwedbankPay/Api/Service/Trustly/Resource/Request/PaymentPayeeInfo.php  1
- src/SwedbankPay/Api/Service/Trustly/Resource/Request/Payment.php  2
- src/SwedbankPay/Api/Service/Trustly/Resource/Request/PaymentObject.php  1
- tests/TrustlyPaymentTest.php  2
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/CreateSale.php  1
- src/SwedbankPay/Api/Service/Trustly/Request/GetPayment.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetSales.php  1
- src/SwedbankPay/Api/Service/Trustly/Request/Purchase.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetSale.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/CreateReversal.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetTransaction.php  1
         

Complexity decreasing per file
==============================
+ src/SwedbankPay/Api/Service/Paymentorder/Resource/PaymentorderUrl.php  -1
+ src/SwedbankPay/Api/Service/Creditcard/Resource/Request/PaymentUrl.php  -1
         

Clones added
============
- src/SwedbankPay/Api/Service/Trustly/Transaction/Resource/Response/Sale.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/SwedbankPay/swedbank-pay-sdk-php/pullRequest?prid=6066095&bid=19866555)

**Comment by payex-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetTransaction.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetSales.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetReversal.php  1
- src/SwedbankPay/Api/Service/Trustly/Resource/Request/PaymentPayeeInfo.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetTransactions.php  1
- src/SwedbankPay/Api/Service/Trustly/Resource/Request/PaymentObject.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/CreateSale.php  1
- src/SwedbankPay/Api/Service/Trustly/Resource/Request/Payment.php  2
- src/SwedbankPay/Api/Service/Trustly/Request/Purchase.php  1
- tests/TrustlyPaymentTest.php  2
- src/SwedbankPay/Api/Service/Trustly/Transaction/Resource/Response/Sale.php  1
- src/SwedbankPay/Api/Service/Trustly/Request/GetPayment.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetSale.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetReversals.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/CreateReversal.php  1
         

Complexity decreasing per file
==============================
+ src/SwedbankPay/Api/Service/Creditcard/Resource/Request/PaymentUrl.php  -1
+ src/SwedbankPay/Api/Service/Paymentorder/Resource/PaymentorderUrl.php  -1
         

Clones added
============
- tests/TrustlyPaymentTest.php  2
- src/SwedbankPay/Api/Service/Trustly/Transaction/Resource/Response/Sale.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/SwedbankPay/swedbank-pay-sdk-php/pullRequest?prid=6066095&bid=19866555)

**Comment by payex-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetSales.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetTransactions.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetTransaction.php  1
- src/SwedbankPay/Api/Service/Trustly/Request/Purchase.php  1
- src/SwedbankPay/Api/Service/Trustly/Request/GetPayment.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/CreateSale.php  1
- src/SwedbankPay/Api/Service/Trustly/Resource/Request/PaymentObject.php  1
- tests/TrustlyPaymentTest.php  2
- src/SwedbankPay/Api/Service/Trustly/Transaction/Resource/Response/Sale.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetSale.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetReversals.php  1
- src/SwedbankPay/Api/Service/Trustly/Resource/Request/Payment.php  2
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/GetReversal.php  1
- src/SwedbankPay/Api/Service/Trustly/Resource/Request/PaymentPayeeInfo.php  1
- src/SwedbankPay/Api/Service/Trustly/Transaction/Request/CreateReversal.php  1
         

Complexity decreasing per file
==============================
+ src/SwedbankPay/Api/Service/Creditcard/Resource/Request/PaymentUrl.php  -1
+ src/SwedbankPay/Api/Service/MobilePay/Resource/Request/PaymentUrl.php  -1
+ src/SwedbankPay/Api/Service/Paymentorder/Resource/PaymentorderUrl.php  -1
+ src/SwedbankPay/Api/Service/Vipps/Resource/Request/PaymentUrl.php  -1
         

Clones added
============
- tests/TrustlyPaymentTest.php  2
- src/SwedbankPay/Api/Service/Trustly/Transaction/Resource/Response/Sale.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/SwedbankPay/swedbank-pay-sdk-php/pullRequest?prid=6066095&bid=19866555)

**Pull request merged**